plugins/aver-hid/fu-aver-safeisp-device.c

```diff
@@ -0,0 +1,448 @@
+	cx3_fw_buf = g_bytes_get_data(cx3_fw, &cx3_fw_size);
+	if (cx3_fw_size > 256 * 1024)
+		return FALSE;
+	/* calculate CX3 firmware checksum */
+	for (guint i = 0; i < cx3_fw_size; i++)
+		cx3_checksum += cx3_fw_buf[i];
+	/* M12 fw file size should be less than 3MB */
+	m12_fw_buf = g_bytes_get_data(m12_fw, &m12_fw_size);
+	if (m12_fw_size > 3 * 1024 * 1024)
+		return FALSE;
+	/* calculate M12 firmware checksum */
+	for (guint i = 0; i < m12_fw_size; i++)
+		m12_checksum += m12_fw_buf[i];
+
+	/* Check if the device supports safeisp */
+	if (!fu_aver_safeisp_device_support(self, error))
+		return FALSE;
+
+	/* CX3 safeisp prepare */
+	if (!fu_aver_safeisp_device_upload_prepare(self, ISP_CX3, cx3_fw_size, error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* CX3 safeisp erase flash */
+	if (!fu_aver_safeisp_device_erase_flash(self, ISP_CX3, 0x0, error))
+		return FALSE;
+
+	/* CX3 safeisp firmware upload */
+	chunks = fu_chunk_array_new_from_bytes(cx3_fw, 0x00, 512);
+	if (!fu_aver_safeisp_device_upload(self,
+					   chunks,
+					   fu_progress_get_child(progress),
+					   ISP_CX3,
+					   error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* CX3 safeisp checksum */
+	if (!fu_aver_safeisp_device_upload_checksum(self, ISP_CX3, cx3_checksum, error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* M12 safeisp prepare */
+	if (!fu_aver_safeisp_device_upload_prepare(self, ISP_M12, m12_fw_size, error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* M12 safeisp erase flash */
+	if (!fu_aver_safeisp_device_erase_flash(self, ISP_M12, 0x0, error))
+		return FALSE;
+
+	/* M12 safeisp firmware upload */
+	chunks = fu_chunk_array_new_from_bytes(m12_fw, 0x00, 512);
+	if (!fu_aver_safeisp_device_upload(self,
+					   chunks,
+					   fu_progress_get_child(progress),
+					   ISP_M12,
+					   error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* M12 safeisp checksum */
+	if (!fu_aver_safeisp_device_upload_checksum(self, ISP_M12, m12_checksum, error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* update device */
+	if (!fu_aver_safeisp_device_update(self, ((1 << ISP_CX3) | (1 << ISP_M12)), 0x0, error))
+		return FALSE;
+
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+
+	/* success! */
+	return TRUE;
+}
+
+static void
+fu_aver_safeisp_device_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 0, "detach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 68, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 31, "attach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 1, "reload");
+}
+
+static void
+fu_aver_safeisp_device_init(FuAverSafeispDevice *self)
+{
+	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_QUAD);
+	fu_device_add_protocol(FU_DEVICE(self), "com.aver.hid");
```

Comment:
is it the same protocol? I'd argue it's a new one -- maybe `com.aver.safeisp`

Reply:
We will add a new protocol, `com.aver.safeisp`, and earnestly request its inclusion in the protocol options of LVFS. Thank you.

Reply:
Added!